### ISSUE TITLE
With kernel 4.4.0, "mount -o loop" hangs against a file in a 'direct_io' fuse mount

### ISSUE THREAD

**Issue by skipca** [libfuse/libfuse#56]
I am using the handle-based example that comes with the FUSE download.

_This problem manifests in kernel 4.4.0. No such problem is seen in kernel 3.16.0_

Trying to create a loop mount for a disk image file in a fuse mount hangs if the fuse mount was created with the 'direct_io' option:

```
# mount fuse WITH 'direct_io'
$ example/fusexmp_fh -o direct_io /tmp/fusexmp/fuse-exmp-mnt

# trying to loop mount an image file hangs
$ mount -o loop /tmp/fuse-exmp-mnt/share_disk.img /tmp/fuse-looptest
<hangs>

$ ps -elf | grep mount
0 D root     17114  3767  0  80   0 -  4676 lock_p 12:25 pts/1    00:00:00 mount -o loop /tmp/fuse-exmp-mnt/share_disk.img /tmp/fuse-looptest
```

No hang is seen (even in kernel 4.4.0) if 'direct_io' isn't specified for the fuse mount:

```
# mount fuse WITHOUT 'direct_io'
$ example/fusexmp_fh /tmp/fusexmp/fuse-exmp-mnt

# trying to loop mount an image file works this time
$ mount -o loop /tmp/fuse-exmp-mnt/share_disk.img /tmp/fuse-looptest
<works>
```

I am using kernel 4.4.0-21-generic and fuse library 2.9.0.

/var/log/messages has call traces fingering fuse_direct_io() and fuse_release_user_pages():

```
Jun 16 12:09:51 Node001 kernel: [ 1200.163685] loop0           D ffff8800aa90b8e8     0  3935      2 0x00000000
Jun 16 12:09:51 Node001 kernel: [ 1200.163691]  ffff8800aa90b8e8 ffffffff81e12540 ffff8800aab08000 ffff8800aa90b908
Jun 16 12:09:51 Node001 kernel: [ 1200.163694]  ffff8800aa90c000 0000000000000000 7fffffffffffffff 0000000000000002
Jun 16 12:09:51 Node001 kernel: [ 1200.163697]  ffffffff81806970 ffff8800aa90b908 ffffffff81806289 ffff8800aa90b958
Jun 16 12:09:51 Node001 kernel: [ 1200.163700] Call Trace:
Jun 16 12:09:51 Node001 kernel: [ 1200.163713]  [<ffffffff81806970>] ? out_of_line_wait_on_atomic_t+0xd0/0xd0
Jun 16 12:09:51 Node001 kernel: [ 1200.163717]  [<ffffffff81806289>] schedule+0x39/0x80
Jun 16 12:09:51 Node001 kernel: [ 1200.163720]  [<ffffffff81808c82>] schedule_timeout+0x222/0x2b0
Jun 16 12:09:51 Node001 kernel: [ 1200.163726]  [<ffffffff8102d65b>] ? __switch_to+0x1bb/0x570
Jun 16 12:09:51 Node001 kernel: [ 1200.163732]  [<ffffffff810f0a85>] ? ktime_get+0x35/0xa0
Jun 16 12:09:51 Node001 kernel: [ 1200.163736]  [<ffffffff81806970>] ? out_of_line_wait_on_atomic_t+0xd0/0xd0
Jun 16 12:09:51 Node001 kernel: [ 1200.163739]  [<ffffffff81805787>] io_schedule_timeout+0xa7/0x110
Jun 16 12:09:51 Node001 kernel: [ 1200.163742]  [<ffffffff8180698f>] bit_wait_io+0x1f/0x60
Jun 16 12:09:51 Node001 kernel: [ 1200.163745]  [<ffffffff81806591>] __wait_on_bit_lock+0x71/0xb0
Jun 16 12:09:51 Node001 kernel: [ 1200.163751]  [<ffffffff811871ee>] __lock_page+0xae/0xb0
Jun 16 12:09:51 Node001 kernel: [ 1200.163757]  [<ffffffff810c0800>] ? wake_atomic_t_function+0x30/0x30    
Jun 16 12:09:51 Node001 kernel: [ 1200.163761]  [<ffffffff81191d32>] set_page_dirty_lock+0x42/0x50
Jun 16 12:09:51 Node001 kernel: [ 1200.163766]  [<ffffffff8130c0a0>] fuse_release_user_pages.isra.22+0x50/0x70
Jun 16 12:09:51 Node001 kernel: [ 1200.163770]  [<ffffffff8130e783>] fuse_direct_io+0x3b3/0x590
Jun 16 12:09:51 Node001 kernel: [ 1200.163773]  [<ffffffff8130e9b5>] __fuse_direct_read+0x55/0x80
Jun 16 12:09:51 Node001 kernel: [ 1200.163776]  [<ffffffff8130ea19>] fuse_direct_read_iter+0x39/0x40
Jun 16 12:09:51 Node001 kernel: [ 1200.163781]  [<ffffffff812047e2>] vfs_iter_read+0x62/0xa0
Jun 16 12:09:51 Node001 kernel: [ 1200.163787]  [<ffffffff8155c7e5>] lo_read_simple.isra.26+0xb5/0x1f0
Jun 16 12:09:51 Node001 kernel: [ 1200.163791]  [<ffffffff8155d8ed>] loop_queue_work+0x51d/0x730
Jun 16 12:09:51 Node001 kernel: [ 1200.163794]  [<ffffffff810b7cd6>] ? put_prev_task_fair+0x26/0x40
Jun 16 12:09:51 Node001 kernel: [ 1200.163798]  [<ffffffff81805b3d>] ? __schedule+0x34d/0x9b1
Jun 16 12:09:51 Node001 kernel: [ 1200.163802]  [<ffffffff8109e509>] kthread_worker_fn+0x79/0x160
Jun 16 12:09:51 Node001 kernel: [ 1200.163805]  [<ffffffff8109e490>] ? flush_kthread_worker+0x70/0x70
Jun 16 12:09:51 Node001 kernel: [ 1200.163807]  [<ffffffff8109e339>] kthread+0xc9/0xe0
Jun 16 12:09:51 Node001 kernel: [ 1200.163810]  [<ffffffff8109e270>] ? kthread_stop+0x100/0x100
Jun 16 12:09:51 Node001 kernel: [ 1200.163813]  [<ffffffff81809e4f>] ret_from_fork+0x3f/0x70
Jun 16 12:09:51 Node001 kernel: [ 1200.163816]  [<ffffffff8109e270>] ? kthread_stop+0x100/0x100
```

Any help on this issue would be appreciated.

Thanks,

**Comment by Nikratio** [comment]
Thanks for the report!

As far as I can tell, this is a problem with *FUSE*, the Linux kernel module. This GitHub project is primarily for *libfuse*, the userspace library. I'm happy to keep track of kernel issues in this  issue tracker as well, but please note that it's quite unlikely that reporting kernel issues here will be sufficient to get them fixed. 

The FUSE kernel module is maintained in the mainline linux kernel, so the best way to get issues addressed is to raise them on the Linux kernel mailing lists, e.g. [linux-kernel](http://vger.kernel.org/vger-lists.html#linux-kernel) or [linux-fsdevel](http://vger.kernel.org/vger-lists.html#linux-fsdevel).

**Comment by skipca** [comment]
Ah, sorry about that, will post to the kernel mailing lists. Thanks.

**Comment by Nikratio** [comment]
Do you have an update on this? If this has been resolved in the kernel, I'd like to close this issue. Thanks!

**Comment by birdie-github** [comment]
This will be fixed in 4.8.0 and perhaps in later 4.7.x releases:

https://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/commit/fs/fuse?id=8fba54aebbdf1f999738121922e74bf796ad60ee

**Comment by Nikratio** [comment]
Thanks for the update! I'll close this issue then.
